doc/dev/debugging.rst

```diff
@@ -118,98 +118,52 @@ Will print out something close to the following:
-For example, let's imagine that a user is interested in testing different optimization levels when compiling LLVM IR to binary using ``llc``.
-The user would then create a ``PassPipeline`` that replaces the ``LLVMIRToObjectFile`` class.
-First let's take a look at the ``LLVMIRToObjectFile``.
+The compilation steps which take MLIR as an input and lower it to binary are broken into MLIR pass
+pipelines.  The ``pipeline`` argument of the ``qjit`` function may be used to alter the steps used
+for compilation. The default set of pipelines is defined via ``catalyst.compiler.DEFAULT_PIPELINES``
+list. Its structure is shown below.
 
 .. code-block:: python
 
-    class LLVMIRToObjectFile(PassPipeline):
-        """LLVMIR To Object File."""
-    
-        _executable = get_executable_path("llvm", "llc")
-        _default_flags = [
-            "--filetype=obj",
-            "--relocation-model=pic",
-        ]
-    
-        @staticmethod
-        def get_output_filename(infile):
-            path = pathlib.Path(infile)
-            if not path.exists():
-                raise FileNotFoundError("Cannot find {infile}.")
-            return str(path.with_suffix(".o"))
-
-
-The ``LLVMDialectTOLLVMIR`` and all classes derived from ``PassPipeline`` must define an ``_executable`` and ``_default_flags`` fields.
-The ``_executable`` field is string that corresponds to the command that will be used to execute in a subprocess.
-The ``_default_flags`` are the flags that will be used when running the executable.
-The method ``get_output_filename`` computes the name of the output file given an input file.
-It is expected that the output of a ``PassPipeline`` will be fed as an input to the following ``PassPipeline``.
-
-From here, we can see that in order for the user to test different optimization levels, all that is needed is create a class that extends either ``PassPipeline`` or ``LLVMDialectToLLVMIR`` and appends the ``-O3`` flag to the ``_default_flags`` field. For example, either of the following classes would work:
-
-
-.. code-block:: python
-
-    class MyLLCOpt(PassPipeline):
-        """LLVMIR To Object File."""
-    
-        _executable = get_executable_path("llvm", "llc")
-        _default_flags = [
-            "--filetype=obj",
-            "--relocation-model=pic",
-            "-O3",
-        ]
-    
-        @staticmethod
-        def get_output_filename(infile):
-            path = pathlib.Path(infile)
-            if not path.exists():
-                raise FileNotFoundError("Cannot find {infile}.")
-            return str(path.with_suffix(".o"))
-
-or
-
-.. code-block:: python
-
-    class MyLLCOpt(LLVMIRToObjectFile):
-        """LLVMIR To Object File."""
-    
-        _default_flags = [
-            "--filetype=obj",
-            "--relocation-model=pic",
-            "-O3",
+    DEFAULT_PIPELINES = [
+        (
+            "HLOLoweringPass",
+            [
+                "canonicalize",
+                "func.func(chlo-legalize-to-hlo)",
```

Comment:
I don't really the second part here with "invoke as calls from Python" since these will never be invoked directly from Python. I'm not sure how much notes we need besides `func.func()`. What note do you have in mind?

Reply:
What's the reasoning behind `func.func` if someone asks? That's the note I had in mind for this part, but this may be so obvious (which wasn't for me at the first look) :smile:

Reply:
`func.func` is called an [operation anchor.](https://mlir.llvm.org/docs/PassManagement/#textual-pass-pipeline-specification) Essentially transformations may be defined on specific operations. An operation anchor tells MLIR over which operations to apply the transform.